### PR TITLE
Emacs integration with `dune --watch`

### PR DESCRIPTION
**Pull request by Gopiandcode**
Opening this pull request for comments on an possible extension to the dune-mode emacs package to add support for running dune with the `--watch` flag.

Currently the change adds a file `dune-watch.el` which adds a new minor mode `dune-watch-minor-mode` which starts a dune subprocess in the background and pops up a buffer when the compile output contains an error.

This seemed like an appropriate place to submit this, if not, I can package it as an independent melpa package.

Signed-off-by: Kiran Gopinathan <kirang@comp.nus.edu.sg>

### CODE REVIEW

**Comment by rgrinberg** [comment]
Note that we plan to integrate dune-watch with all editors via an upcoming rpc mechanism and lsp diagnostics. In practice, this would mean that build errors would show up inside flycheck/flymake (depending on your lsp client) in Emacs.

However, if people find your extension useful, I see no problem with including it.

**Comment by Gopiandcode** [comment]
Sure, if there's plans to add this functionality coming soon, then I guess this would be unnecessary and I can understand leaving it out. 

As for whether people find this useful, my only evidence is the associated ocaml discuss thread on it: https://discuss.ocaml.org/t/how-to-build-the-dune-based-project-from-inside-the-emacs/4263

which may or may not be compelling enough.

**Comment by rgrinberg** [comment]
> Sure, if there's plans to add this functionality coming soon, then I guess this would be unnecessary and I can understand leaving it out

It's not going to be completely analogous to your minor mode, and will only be usable with lsp. So I still think this is a useful contribution.

File: editor-integration/emacs/dune-watch.el [hunk 1]
```diff
@@ -0,0 +1,196 @@
+;;; dune-watch.el --- provides integration with dune watch  -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2021  Kiran Gopinathan
+
+;; Author: Kiran Gopinathan <kirang@comp.nus.edu.sg>
+;; Keywords: 
+
+;; This program is free software; you can redistribute it and/or modify
```
Comment (rgrinberg) [inline editor-integration/emacs/dune-watch.el:8]: Do you mind changing this to MIT to match the rest of dune?

Comment (Gopiandcode) [reply editor-integration/emacs/dune-watch.el]: Sure, done.

**Comment by rgrinberg** [comment]
Thanks!

**Comment by jberdine** [comment]
Maybe it's just me, but emacs' `next-error` function jumps to the error reported in the `*dune-watch*` buffer as expected immediately after the minor mode is enabled, but after saving a file and thereby triggering a rebuild, the errors shown in the refreshed `*dune-watch*` buffer are not highlighted and not jumped to.

**Comment by Gopiandcode** [comment]
That's strange - I think it may be due to the fact that after a rebuild, the error buffer is cleared, but whatever initialization that compilation-mode uses is not rerun - I'll have a look into it. Generally, I just use merlin's `merlin-error-next` to jump to errors, so I hadn't run into this.

**Comment by Gopiandcode** [comment]
Hey @roddyyaga, there are two ways you might be able to solve your problem - either:
 1. disabling color output by dune (I believe it uses the `TERM` variable (see [here](https://github.com/ocaml/dune/blob/d75bd885262b98075299e9145a2aa3bd1a54a35d/otherlibs/stdune-unstable/ansi_color.ml#L126) - setting it to dumb in your command invocation should disable colors - `TERM=dumb opam exec -- dune build --watch`)
 2. Use the emacs package `ansi-color` to update your compilation buffer to actually understand the escape codes (I would recommend this - there are probably some guides online for setting it up with spacemacs).

**Pull request merged**